keps/sig-node/3545-improved-multi-numa-alignment/README.md

```diff
@@ -0,0 +1,444 @@
+
+Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
+This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.
+
+To summarize properties of `prefer-closest-numa-nodes` policy:
+
+* Preserve all properties of `best-effort` and `restricted` policies
+* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.
+
+When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
+Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).
+
+We would need to extend the `MachineInfo` struct with Distances field which would be an array of uint64 which will describe the distance between a given node and other nodes in the system.
+This information can be read from sysfs:
+
+```go 
+const     distanceFile = "distance"
+
+func (fs *realSysFs) GetDistance(nodePath string) (string, error) {
+    distancePath := fmt.Sprintf("%s/%s", nodePath, distanceFile)
+    distance, err := ioutil.ReadFile(distancePath)
+    if err != nil {
+        return "", err
+    }
+    return strings.TrimSpace(string(distance)), err
+}
+
+func getDistance(sysFs sysfs.SysFs, nodeDir string) ([]uint64, error) {
+    rawDistance, err := sysFs.GetDistance(nodeDir)
+    if err != nil {
+        //Ignore if per-node info is not available.
+        klog.Warningf("Found node without distance information, nodeDir: %s", nodeDir)
+        return nil, nil
+    }
+
+    distances := []uint64{}
+    for _, distance := range strings.Split(rawDistance, " ") {
+        distanceInt, err := strconv.ParseUint(distance, 10, 64)
+        if err != nil {
+            return nil, fmt.Errorf("cannot convert %s to int", distance)
+        }
+        distances = append(distances, distanceInt)
+    }
+
+    return distances, nil
+}
+```
+
+But before this lands into `cadvisor`, this information would have to be discovered by `kubelet`.
+
+### Implementation strategy
+
+- Introduce new flag in Kubelet called `topology-manager-policy-options`, which when specified with `prefer-closest-numa-nodes` will modify the behavior of `best-effort` and `restricted` policy to pick NUMA nodes based on average distance between them.
+- The `TopologyManagerPolicyOptions` flag is propagated to `ContainerManager` and later to `TopologyManager`.
+- When `TopologyManager` is being created it discovers distances between NUMA nodes and stores them inside `manager` struct. This is temporary until `distance` information will land into `cadvisor`.
```

Comment:
How this temporary code path will be handled and how the phase off will be handled? are we looking for this temporary approach because there is uncertainity about the timeline of merging the new feature in cadvisor or there are other factors?
in general, I'd prefer to try very hard to avoid this temporary solution, because temporary solutions tend to stick around very long

Reply:
There is precedence for this with the existence of the `GetNUMANode()` function we had in the original `TopologyManager` code while we waited for `cadvisor` to catch up with the info it supplied in its `MachineInfo` struct.

`GetNUMANode()`  was introduced here:
https://github.com/kubernetes/kubernetes/commit/ecc14fe661c22f5da967a7ff50cfb3aead60905b

And removed here:
https://github.com/kubernetes/kubernetes/commit/a047e8aa1b705bb7e5be881fb63cf90a218b60d0

@knopt, we should point at this in the KEP as an example of how this has been done in the past

Reply:
I'm still not 100% happy with temporary approach, especially if enhancing cadvisor is compatible with the timelines we're looking at, but if there is consensus this approach is OK - and having a precedent surely helps - I'm *NOT* gonna block on this point here.

Reply:
Thanks for the comments, I've updated both implementation plan and graduation criteria. I've set cadvisor update to be blocker for alpha to beta graduation, and included the links to `GetNUMANode()` introduction and removal.